scenario/errors.py

```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Exceptions raised by the framework.
+
+Note that these exceptions are not meant to be caught by charm authors. They are
+used by the framework to signal errors or inconsistencies in the charm tests
+themselves.
+"""
+
+
+class ContextSetupError(RuntimeError):
+    """Raised by Context when setup fails."""
+
+
+class AlreadyEmittedError(RuntimeError):
+    """Raised when ``run()`` is called more than once."""
+
+
+class ScenarioRuntimeError(RuntimeError):
+    """Base class for exceptions raised by the runtime module."""
+
+
+class UncaughtCharmError(ScenarioRuntimeError):
```

Comment:
Discussion topic: should all public exceptions be re-exported in `__init__.py` ?

Or should a bunch of these exceptions be private?

Reply:
The intention here is that the top-level `scenario` namespace has everything that you need to write tests, including type annotations. It'll also be what's in `ops.testing` when we let you do `pip install ops[testing]` to get Scenario.

All of the classes in `scenario.errors` are ones that Scenario will raise when the tests themselves are broken in some way - no-one should ever be catching them. They exist because it's nicer to see a custom named exception than `RuntimeError` or `Exception` or similar with a message that needs to be read to understand what you've done incorrectly. They're public because you should be able to look up what they mean, but they aren't part of the toolkit for writing tests.